### DISCUSSION TITLE
Adding a kernel with two or more variables

### DISCUSSION THREAD

**Discussion by iampham** [idaholab/moose#17524] [Q&A General] [answered]
Hello,

I've been trying to implement the following expression into my residual:

![image](https://user-images.githubusercontent.com/23412680/113798977-35123500-9722-11eb-8781-172804a24042.png)

where phi_i is the main "variable" in the sense of what you'd put in a default MOOSE kernel. However, I want to find the variational derivative of a function with respect to another variable phi_j, and add the term into the same residual that acts on the variable phi_i. These parameters are individually defined in the Variables system.

Adding the variational derivative of F with respect to phi_i is already implemented in MOOSE's PhaseField module (AllenCahn.c). However, I believe I need to create my own file to add the variational derivative of F with respect to phi_j into the residual. This process doesn't seem trivial to me because I'm not familiar with what the constructors for a kernel do when they take variables as input. 

I have two questions regarding this process.
1. Does MOOSE create a unique residual for each variable defined in the variables system? Or do they get added into some global residual?
2. How can I add an additional variable as input into the kernel? If I do this, how can I refer to it in the .C file? For instance, a `variable` in a kernel usually gets referred to as `_var.name()`. If I add a second variable `variable2`, is there a get function that I can easily use?

I've been poring through the documentation for an answer, but this seems like enough of an edge case to make it difficult to find. If you have a link to the relevant documentation, please let me know it!

Thank you for your help!

EDIT: Just in case anybody stumbles across this problem, I found an additional useful page that describes how MOOSE variables work: https://mooseframework.inl.gov/source/variables/MooseVariableBase.html

**Comment by YaqiWang** [comment, marked as answer]
I will try answering your questions:

1. residual vector is composed of parts of number of variables, each for a `primal` (with respect to auxiliary) variable.
2. normal kernel can only evaluate residual contributions on local elements for one variable, specified by `variable` parameter. If you want a kernel to operate on multiple variables, you can consider array variable/kernel (documented here: https://mooseframework.inl.gov/source/variables/ArrayMooseVariable.html).

Hope these help.

**Reply by iampham** [reply]
Thank you for the answer! I took a look at the documentation and I think this is what I had in mind.

**Reply by dzheng26** [reply]
It is good that I find this discussion. I'm thinking about implemeningt a tensor mechanics kernel. In my case, I need `disp_x, disp_y and disp_z` as a vector together. I had a look of the current tensor mechanics kernel `StressDivergenceTensors.C`. It basically handles `disp_x, disp_y and disp_z` separately. The function `computeOffDiagJacobian` is used 9 times for each element for the combination of 3 variables `disp_x, disp_y, and disp_z`. It seems to be not the optimal way to do it and it also does not fit my need.

With the new Arraykernel, it seems to be a better kernel type for tensor mechanics. I should be able to implement my case. Is this the intension of this type of kernel? @YaqiWang

**Reply by YaqiWang** [reply]
We developed ArrayKernel for operating array variables that can potentially have hundreds of components. But I think it can find usages for with a few or several components. It can make relevant codes organized together thus easier to maintain. Even with a small number of components, we do not see performance degradation. So I think you can go ahead.

**Reply by dzheng26** [reply]
Another questions: What is the difference between ArrayKernel and VectorKernel?  Is the `variable` used by them be used by things like BCs, Pressure, etc.

**Reply by YaqiWang** [reply]
Array kernels require array variables whose components must be in the same family and order. Vector kernels require a vector variable in a specific type with number of components being space dimension. I think vector kernels are designed for some specific problems, @lindsayad can possibly answer this better. We have BCs, kernels, etc. for both array and vector variables.

**Reply by jiangwen84** [reply]
@dzheng-conti I think you should use `VectorKernel` for vector version of`StressDivergenceTensors`  because that is supposed to work with `VectorVariable`.  You can take a look at `INSADMomentumViscous` on how `VectorKernel` can be used.

**Reply by dzheng26** [reply]
Thanks a lot! @YaqiWang  @jiangwen84

**Reply by dzheng26** [reply]
I had a look of the `VectorKernal` and `ArrayKernal` examples and I had another questions need your help.
My question is that can I use one component of a Vector or Array variable for the scalerKernel (scalerClass).

In tensormechanics module, there are many Classes defined for things like material models, pressure loads, contact etc. They all expect scaler (LAGRANGE) variables, say `disp_x`, `disp_y`, and `disp_z`. If I write a VectorKernel for  StressVivergenceTensors, it expects a vector (LAGRANGE_VEC) variable, say `disp`. If I want to use the already defined Classes in combination of my new VectorKernel, how can I let Moose know that `disp_x`, `disp_y`, and `disp_z` are the three components of `disp`?

**Reply by lindsayad** [reply]
I'd have to see your weak form to really understand what you want to do. But you can always create a vector in your `computeQpResidual` statement from scalar components like so:
```
RealVectorValue disp_vec(_disp_x[_qp], _disp_y[_qp], _disp_z[_qp]);
```

**Reply by dzheng26** [reply]
@lindsayad. Thanks for the reply. I will try to explain what I mean with an example.
Currently in TensorMechanics module, all the classes are using standard variables. To do a simulation with gravity load for example, the input file is something like (I show the input file without using action)
```
[Variables]
  [./disp_x]
    family = LAGRANGE
  [../]
  [./disp_y]
    family = LAGRANGE
  [../]
  [./disp_z]
    family = LAGRANGE
  [../]
[]
[Kernels]
  [elements_x]
    type = StressDivergenceTensors
    component = 0
    variable = disp_x
  []
  [elements_y]
    type = StressDivergenceTensors
    component = 1
    variable = disp_y
  []
  [elements_z]
    type = StressDivergenceTensors
    component = 2
    variable = disp_z
  []
  [gravity_z]
    type = Gravity
    variable = disp_z
    value = -1.0e-3
    function = load
  []
[]
``` 
Three standard variables are defined to representing the displacement vector. The variables are then used for `StressDivergenceTensor` and for `Gravity`. For one element, the `computerOffDiagJacobian` of `StressDivergenceTensor` is used 9 times (loops i_var=disp_x, disp_y, disp_z and j_var=disp_x, disp_y, disp_z) to compute the full jacobian. In other words, for Hex8 element, the full jacobian is a 24x24 matrix. Each call of  `computerOffDiagJacobian`  generates only an 8x8 submatrix of the full jacobian.

For my case,  I cannot compute the 8x8 submatrix for disp_x if disp_y and disp_z is not available. I need disp_x, disp_y and disp_z together so that I can calculate the 24x24 full jacobian in one call of `computeJacobian`.  I thought that I can write a `VectorStressDivergenceTensor`.  Then the input file can become something like
```
[Variables]
  [./dispvec]
    family = LAGRANGE_VEC
  [../]
[]

[Kernels]
  [elements]
    type = VectorStressDivergenceTensors
    variable = dispvec
  []
[]
```
The full jacobian can be calculated with one call of `computeJacobian` since all three components are represented by `dispvec`

My problem come from the things like the `Gravity` kernel. There, `variable=` expects a `standard variable` representing displacement in one (say z) direction. With vector kernel, my displacement is defined as `vector variable`. Is there a way to pass one component of the vector variable `dispvec` to the Gravity kernel which needs a standard variable?
```
[Variables]
  [./dispvec]
    family = LAGRANGE_VEC
  [../]
[]
[Kernels] 
  ...
  [gravity_z]
    type = Gravity
    variable = ? ## what should be here?
    value = -1.0e-3
    function = load
  []
[]
```
Without this function, it seems to me that I cannot use a vector kernel in combination of other classes in the TensorMechanics module which use standard variable. There are many of them (Gravity, Pressure, Contact, etc.) already defined. It is a lot of work if we have to rewrite them to use vector variable.

In fact, it should be also fine if we can use 3 standard variables instead of a vector variable for VectorKernel. Libmesh example "examples/systems_of_equations/systems_of_equations_ex7" seems did it this way.

Thanks for your help.

**Reply by lindsayad** [reply]
It is not possible to make Gravity (or others like it) use a vector variable as it stands now. We could possibly make it work if we added another parameter to those kernels like ‘component’. But without that additional parameter or rewriting the kernel to genuinely use a vector variable it will be impossible

**Reply by dzheng26** [reply]
@lindsayad, Thanks for your reply. It's sorry to hear that Moose vector variable cannot be used with classes require standard (scaler) variable right now. It is a lot of work to add a new parameter to all those classes since it means change all of them. On the other hand, enable this feature will give Moose a lot of flexibility. I have a suggestion which maybe done without too much work.

For a Moose vector variable `vecvar`, The names `vecvar_x`, `vecvar_y`, and `vecvar_z` can be used to map one component of the vector variable to standard variable so that they can be used the classes requiring standard variable.

The reasons:
1) No interfaces changes for all existing classes.
2) I assume that handling of variable is done in a few class/functions, it can be realized by changing them.
3) Actually, I had a look of the example `./test/tests/kernels/vector_fe/lagrange_vec.i`. There `u` is defined as a vector variable. In the output file, it is saved as two scalers `u_x` and `u_y` for its two components. In a sense, Moose (or libmesh) is already doing this type of mapping. Why not make it consistent. 

I greatly appreciate your opinion. Thanks.

**Reply by lindsayad** [reply]
This maybe be possible as long as you make the shape function ordering match between the vector variable and mixture of standard variables. You are welcome to pursue this idea if you wish. We always appreciate community pull requests. We do not have money for this development ourselves right now though

**Reply by lindsayad** [reply]
You would probably have to make a derivative class of MooseVariableFE<RealVectorValue> to make this work.

**Reply by lindsayad** [reply]
Whether you want to work on it or have us work on it down the road you should create an issue for your idea.

**Reply by dzheng26** [reply]
@lindsayad , Thanks help the help. I will make an issue the the tracker. I will also have a look of MooseVariableFE to see if I can do it myself or ask someone with better C++ to do it for me. Thanks again.
